modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C

```diff
@@ -0,0 +1,627 @@
+                   " and qp ",
+                   _qp,
+                   "\n");
+#endif
+    }
+    iteration++;
+  }
+
+  if (iteration == _maxiterg)
+  {
+#ifdef DEBUG
+    mooseWarning("ComputeMultipleCrystalPlasticityStress: Hardness Integration error. Reached the "
+                 "maximum number "
+                 "of iterations to solve for the state variables at element ",
+                 _current_elem->id(),
+                 " and qp ",
+                 _qp,
+                 "\n");
+#endif
+    _error_tolerance = true;
+  }
+}
+
+void
+ComputeMultipleCrystalPlasticityStress::solveStress()
+{
+  unsigned int iteration = 0;
+  RankTwoTensor dpk2;
+  Real rnorm, rnorm0, rnorm_prev;
+
+  // Calculate stress residual
+  calculateResidualAndJacobian();
+  if (_error_tolerance)
+  {
+#ifdef DEBUG
+    mooseWarning("ComputeMultipleCrystalPlasticityStress: Slip increment exceeds tolerance - "
+                 "Element number ",
+                 _current_elem->id(),
+                 " Gauss point = ",
+                 _qp);
+#endif
+    return;
+  }
+
+  rnorm = _residual_tensor.L2norm();
+  rnorm0 = rnorm;
+
+  // Check for stress residual tolerance; different from user object version which
+  // compares the absolute tolerance of only the original rnorm value
+  while (rnorm > _rtol * rnorm0 && rnorm > _abs_tol && iteration < _maxiter)
+  {
+    // Calculate stress increment
+    dpk2 = -_jacobian.invSymm() * _residual_tensor;
+    _pk2[_qp] = _pk2[_qp] + dpk2;
+
+    calculateResidualAndJacobian();
+
+    if (_error_tolerance)
+    {
+#ifdef DEBUG
+      mooseWarning("ComputeMultipleCrystalPlasticityStress: Slip increment exceeds tolerance - "
+                   "Element number ",
+                   _current_elem->id(),
+                   " Gauss point = ",
+                   _qp);
+#endif
+      return;
+    }
+
+    rnorm_prev = rnorm;
+    rnorm = _residual_tensor.L2norm();
+
+    if (_use_line_search && rnorm > rnorm_prev && !lineSearchUpdate(rnorm_prev, dpk2))
+    {
+#ifdef DEBUG
+      mooseWarning("ComputeMultipleCrystalPlasticityStress: Failed with line search");
+#endif
+      _error_tolerance = true;
+      return;
+    }
+
+    if (_use_line_search)
+      rnorm = _residual_tensor.L2norm();
+
+    iteration++;
+  }
+
+  if (iteration >= _maxiter)
+  {
+#ifdef DEBUG
+    mooseWarning("ComputeMultipleCrystalPlasticityStress: Stress Integration error rmax = ",
+                 rnorm,
+                 " and the tolerance is ",
+                 _rtol * rnorm0,
+                 "when the rnorm0 value is ",
+                 rnorm0,
+                 "for element ",
+                 _current_elem->id(),
+                 " and qp ",
+                 _qp);
+#endif
+    _error_tolerance = true;
+  }
+}
+
+// Calculates stress residual equation and jacobian
+void
+ComputeMultipleCrystalPlasticityStress::calculateResidualAndJacobian()
+{
+  calcResidual();
+  if (_error_tolerance)
+    return;
+  calcJacobian();
```

Comment:
Why is the Jacobian calculation skipped if the `_error_tolerance` flag is set?

Reply:
Oh, because at that point we'd increase the number of substeps

Reply:
perhaps a name, like `_internal_convergence_failed` or something like that would make more sense